modules/rostests/kmtests/ntos_ob/ObInsert.c

```diff
@@ -0,0 +1,651 @@
+        {
+            /* Reference the existing object type by handle */
+            Status = ObReferenceObjectByHandle(ObjectTypeHandle,
+                                               0,
+                                               NULL,
+                                               KernelMode,
+                                               (PVOID)&DummyType,
+                                               NULL);
+
+            ok_eq_hex(Status, STATUS_SUCCESS);
+
+            if (!skip(Status == STATUS_SUCCESS && DummyType, "No pointer to the dummy object's body\n"))
+            {
+                /* Set the procedures for the existing object type */
+                DummyType->TypeInfo.CloseProcedure = CloseProc;
+                DummyType->TypeInfo.DeleteProcedure = DeleteProc;
+                DummyType->TypeInfo.DumpProcedure = DumpProc;
+                DummyType->TypeInfo.OpenProcedure = OpenProc;
+                DummyType->TypeInfo.ParseProcedure = ParseProc;
+                DummyType->TypeInfo.OkayToCloseProcedure = OkayToCloseProc;
+                DummyType->TypeInfo.QueryNameProcedure = QueryNameProc;
+            }
+
+            /* Close the object type handle */
+            if (ObjectTypeHandle != NULL)
+            {
+                Status = ZwClose(ObjectTypeHandle);
+            }
+        }
+    }
+
+    ok_eq_hex(Status, STATUS_SUCCESS);
+}
+
+/*!
+ * @brief This function cleans up resources that were not cleaned up during the tests.
+ */
+static
+VOID
+ObInsert_Cleanup()
+{
+    NTSTATUS Status;
+
+    /* Make the directory object temporary (which is permanent) if the handle is valid */
+    if (!skip(DirectoryHandle != NULL, "No directory handle\n"))
+    {
+        Status = ZwMakeTemporaryObject(DirectoryHandle);
+        ok_eq_hex(Status, STATUS_SUCCESS);
+
+        /* Close the directory handle */
+        Status = ZwClose(DirectoryHandle);
+        ok_eq_hex(Status, STATUS_SUCCESS);
+    }
+}
+
+/*!
+ * @brief This function tests for successful object insertion.
+ */
+static
+VOID
+ObInsert_Success()
+{
+    NTSTATUS Status;
+    UNICODE_STRING ObjectName;
+    PDUMMY_TYPE Object;
+    HANDLE Handle;
+    POBJECT_HEADER ObjectHeader;
+
+    /* Initialize the object name */
+    RtlInitUnicodeString(&ObjectName,
+                         L"\\ObInsertTest\\DummyObject_Success");
+
+    /* Initialize object attributes with the name */
+    InitializeObjectAttributes(&DummyObjectAttributes,
+                               &ObjectName,
+                               OBJ_CASE_INSENSITIVE,
+                               NULL,
+                               NULL);
+
+    /* Create the object */
+    Status = ObCreateObject(KernelMode,
+                            DummyType,
+                            &DummyObjectAttributes,
+                            KernelMode,
+                            NULL,
+                            sizeof(DUMMY_TYPE),
```

Comment:
```suggestion
                            sizeof(*Object),
```
as "weird" as this may look, there is no dereferencing of the pointer, since the sizeof expression is evaluated at compile time.

Reply:
Thanks!